### PR TITLE
[helper] Support setting properties on schemas and fields.

### LINKED ISSUE #202: [helper] Support addProp().
They are needed at least for `Schema`.

It looks like `Schema.Field` also extends JsonProperties, so we might need to support them in Schema.Field also?

### PR DESCRIPTION
**Pull request by srramach**
Also, revamp the get-props paths to fix bugs and make things more
consistent across versions.

Fixes #202.

### CODE REVIEW

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/Jackson1Utils.java [hunk 1]
```diff
@@ -11,13 +11,22 @@
...
+import org.codehaus.jackson.JsonParser;
 import org.codehaus.jackson.map.ObjectMapper;
 import org.codehaus.jackson.node.JsonNodeFactory;
+import org.codehaus.jackson.node.TextNode;
 
 
 public class Jackson1Utils {
-  static final String BYTES_CHARSET = "ISO-8859-1";
+  private static final String BYTES_CHARSET = "ISO-8859-1";
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
```
Comment (radai-rosenblatt) [inline helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/Jackson1Utils.java:24]: do we want to allow some of the non standard stuff? jackson as used by avro does :-(

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/Jackson1Utils.java [hunk 2]
```diff
@@ -61,4 +70,43 @@ public static JsonNode toJsonNode(Object datum) {
...
+      if (node == null) {
+        // Happens when str.isEmpty().
+        throw new NullPointerException("str cannot be an empty string");
+      }
+      // Calling parser.nextToken() can raise an exception itself, if there are garbage chars at the end.
+      // E.g., when str == "1,2", the ',' is the garbage char (raises JsonParseException).
+      if (parser.nextToken() != null) {
+        // Or, if that succeeds and returns a valid token, the extra chars (while not garbage) are still unwanted.
+        // E.g., when str == "1 2", the '2' is the extra token.
+        throw new IllegalStateException("str cannot have extra trailing chars; must be a single literal");
```
Comment (radai-rosenblatt) [inline helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/Jackson1Utils.java:90]: could you include location (from parser.getTokenLocation()) and maybe the "junk string" ? (the part of the string remaining past the location where parsing stopped)

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/Jackson2Utils.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      if (node == null) {
+        // Happens when str.isEmpty().
+        throw new NullPointerException("value cannot be an empty string");
+      }
+      // Calling parser.nextToken() can raise an exception itself, if there are garbage chars at the end.
+      // E.g., when str == "1,2", the ',' is the garbage char (raises JsonParseException).
+      if (parser.nextToken() != null) {
+        // Or, if that succeeds and returns a valid token, the extra chars (while not garbage) are still unwanted.
+        // E.g., when str == "1 2", the '2' is the extra token.
+        throw new IllegalStateException("value cannot have extra trailing chars; must be a single literal");
```
Comment (radai-rosenblatt) [inline helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/Jackson2Utils.java:40]: same comment as aboce - can get the location() of the current token and/or the "junk" substring

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/StringUtils.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+  /**
+   * @param str a string
+   * @return true if the string starts and ends with double quotes
+   */
+  public static boolean isQuoted(String str) {
+    if (str == null) {
+      return false;
+    }
+    return str.length() >= 2 && str.startsWith("\"") && str.endsWith("\"");
```
Comment (radai-rosenblatt) [inline helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/StringUtils.java:26]: good catch on the length >= 2 :-)

File: helper/impls/helper-impl-110/src/main/java/com/linkedin/avroutil1/compatibility/avro110/Avro110Adapter.java [hunk 5]
```diff
@@ -291,15 +292,31 @@ public SchemaBuilder cloneSchema(Schema schema) {
     }
 
     @Override
-    public String getFieldPropAsJsonString(Schema.Field field, String propName) {
-        Object val = field.getObjectProp(propName);
-        return objectPropToJsonString(val);
+    public String getFieldPropAsJsonString(Schema.Field field, String name) {
+        // Goes from JsonNode to Object to JsonNode (again) to String. Painful, but suffices until somebody complains.
+        JsonNode node = JacksonUtils.toJsonNode(field.getObjectProp(name));
```
Comment (radai-rosenblatt) [inline helper/impls/helper-impl-110/src/main/java/com/linkedin/avroutil1/compatibility/avro110/Avro110Adapter.java:297]: mixed use of jackson 1 and 2 utils ?

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/Jackson1Utils.java (hunk 1 above)
Comment (srramach) [reply helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/Jackson1Utils.java:25]: Sure, happy to add, but I don't know what they are. Could you give some pointers?

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/Jackson1Utils.java (hunk 2 above)
Comment (srramach) [reply helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/Jackson1Utils.java]: Good suggestion! Done.

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/Jackson2Utils.java (hunk 3 above)
Comment (srramach) [reply helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/Jackson2Utils.java]: Done.

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/StringUtils.java (hunk 4 above)
Comment (srramach) [reply helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/StringUtils.java:26]: 😄

**Review by radai-rosenblatt — APPROVED** [review]

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/Jackson1Utils.java (hunk 1 above)
Comment (srramach) [reply helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/Jackson1Utils.java:25]: Radai pointed out offline that Avro's Schema class uses a Jackson mapper configured to allow comments. I think for now, it's okay not to support comments in prop getter/setter methods. We can revisit if somebody runs into an issue.

**Pull request merged**